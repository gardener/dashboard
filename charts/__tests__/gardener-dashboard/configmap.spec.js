#!/usr/bin/env node
//
// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

'use strict'

const yaml = require('js-yaml')
const { omit, pick } = require('lodash')
const { basename } = require('path')
const { helm, helper } = fixtures
const { getCertificate } = helper

const chart = basename(__dirname)
const renderTemplates = (templates, values) => helm.renderChartTemplates(chart, templates, values)

describe('gardener-dashboard', function () {
  describe('configmap', function () {
    const name = 'gardener-dashboard-configmap'
    let templates

    beforeEach(() => {
      templates = [
        'configmap'
      ]
    })

    it('should render the template w/ defaults values', async function () {
      const documents = await renderTemplates(templates, {})
      expect(documents).toHaveLength(1)
      const [configMap] = documents
      expect(omit(configMap, ['data'])).toMatchSnapshot()
      expect(Object.keys(configMap.data)).toEqual(['config.yaml'])
      const config = yaml.safeLoad(configMap.data['config.yaml'])
      expect(config).toMatchSnapshot()
    })

    describe('kubeconfig download', function () {
      it('should render the template w/ `oidc.public`', async function () {
        const values = {
          apiServerCa: getCertificate('apiServerCa'),
          oidc: {
            public: {
              clientId: 'kube-kubectl',
              clientSecret: 'kube-kubectl-secret'
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(omit(config, ['frontend', 'terminal'])).toMatchSnapshot()
      })
    })

    describe('access restrictions', function () {
      it('should render the template w/o `item.options`', async function () {
        const values = {
          frontendConfig: {
            accessRestriction: {
              noItemsText: 'no items text',
              items: [
                {
                  key: 'foo',
                  display: {
                    visibleIf: true
                  },
                  input: {
                    title: 'Foo'
                  }
                }
              ]
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.accessRestriction'])).toMatchSnapshot()
      })

      it('should render the template w/ `item.options`', async function () {
        const values = {
          frontendConfig: {
            accessRestriction: {
              items: [
                {
                  key: 'foo',
                  display: {
                    visibleIf: true,
                    title: 'display Foo',
                    description: 'display Foo description'
                  },
                  input: {
                    title: 'input Foo',
                    description: 'input Foo description',
                    inverted: true
                  },
                  options: [
                    {
                      key: 'foo-option-1',
                      display: {
                        visibleIf: false,
                        title: 'display Foo Option 1',
                        description: 'display Foo  Option 1 description'
                      },
                      input: {
                        title: 'input Foo Option 1',
                        description: 'input Foo  Option 1 description',
                        inverted: false
                      }
                    },
                    {
                      key: 'foo-option-2',
                      display: {
                        visibleIf: true
                      },
                      input: {
                        title: 'input Foo Option 2',
                        description: 'input Foo  Option 2 description',
                        inverted: true
                      }
                    }
                  ]
                }
              ]
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.accessRestriction'])).toMatchSnapshot()
      })
    })

    describe('tickets', function () {
      beforeEach(() => {
        templates.push('secret-github')
      })

      it('should render the template', async function () {
        const values = {
          frontendConfig: {
            ticket: {
              avatarSource: 'gravatar',
              gitHubRepoUrl: 'https://github.com/gardener/tickets',
              hideClustersWithLabels: ['ignore1', 'ignore2'],
              newTicketLabels: ['default-label'],
              issueDescriptionTemplate: 'issue description'
            }
          },
          gitHub: {
            apiUrl: 'https://github.com/api/v3/',
            org: 'gardener',
            repository: 'tickets',
            webhookSecret: 'webhookSecret',
            authentication: {
              username: 'dashboard-tickets',
              token: 'webhookAuthenticationToken'
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(2)
        const [configMap, githubSecret] = documents
        expect(configMap.metadata.name).toBe(name)
        expect(githubSecret.metadata.name).toBe('gardener-dashboard-github')
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.ticket', 'gitHub'])).toMatchSnapshot()
        expect(githubSecret).toMatchSnapshot()
      })
    })

    describe('unreachable seeds', function () {
      it('should render the template', async function () {
        const values = {
          unreachableSeeds: {
            matchLabels: {
              seed: 'unreachable'
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['unreachableSeeds'])).toMatchSnapshot()
      })
    })

    describe('alert', function () {
      it('should render the template w/o `alert.identifier`', async function () {
        const values = {
          frontendConfig: {
            alert: {
              message: 'foo',
              type: 'warning'
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.alert'])).toMatchSnapshot()
      })

      it('should render the template w/ `alert.identifier`', async function () {
        const values = {
          frontendConfig: {
            alert: {
              message: 'foo',
              type: 'warning',
              identifier: 'bar'
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        expect(configMap.metadata.name).toBe(name)
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.alert'])).toMatchSnapshot()
      })
    })

    describe('terminal shortcuts', function () {
      it('should render the template', async function () {
        const values = {
          frontendConfig: {
            terminal: {
              shortcuts: [
                {
                  title: 'title',
                  description: 'description',
                  target: 'foo-target',
                  container: {
                    command: [
                      'command'
                    ],
                    image: 'repo:tag',
                    args: [
                      'a',
                      'b',
                      'c'
                    ]
                  }
                }
              ]
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.terminal.shortcuts'])).toMatchSnapshot()
      })
    })

    describe('themes', function () {
      it('should render the template', async function () {
        const values = {
          frontendConfig: {
            themes: {
              light: {
                primary: '#ff0000',
                'main-navigation-title': 'grey.darken3'
              },
              dark: {
                primary: '#ff0000',
                'main-navigation-title': 'grey.darken3'
              }
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.themes'])).toMatchSnapshot()
      })

      it('should render the template with light theme values only', async function () {
        const values = {
          frontendConfig: {
            themes: {
              light: {
                primary: '#ff0000',
                'main-navigation-title': 'grey.darken3'
              }
            }
          }
        }
        const documents = await renderTemplates(templates, values)
        expect(documents).toHaveLength(1)
        const [configMap] = documents
        const config = yaml.safeLoad(configMap.data['config.yaml'])
        expect(pick(config, ['frontend.themes'])).toMatchSnapshot()
      })
    })
  })
})
