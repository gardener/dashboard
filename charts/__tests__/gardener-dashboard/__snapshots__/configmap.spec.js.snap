// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gardener-dashboard configmap access restrictions should render the template w/ \`item.options\` 1`] = `
Object {
  "frontend": Object {
    "accessRestriction": Object {
      "items": Array [
        Object {
          "display": Object {
            "description": "display Foo description",
            "title": "display Foo",
            "visibleIf": true,
          },
          "input": Object {
            "description": "input Foo description",
            "inverted": true,
            "title": "input Foo",
          },
          "key": "foo",
          "options": Array [
            Object {
              "display": Object {
                "description": "display Foo  Option 1 description",
                "title": "display Foo Option 1",
                "visibleIf": false,
              },
              "input": Object {
                "description": "input Foo  Option 1 description",
                "title": "input Foo Option 1",
              },
              "key": "foo-option-1",
            },
            Object {
              "display": Object {
                "visibleIf": true,
              },
              "input": Object {
                "description": "input Foo  Option 2 description",
                "inverted": true,
                "title": "input Foo Option 2",
              },
              "key": "foo-option-2",
            },
          ],
        },
      ],
    },
  },
}
`;

exports[`gardener-dashboard configmap access restrictions should render the template w/o \`item.options\` 1`] = `
Object {
  "frontend": Object {
    "accessRestriction": Object {
      "items": Array [
        Object {
          "display": Object {
            "visibleIf": true,
          },
          "input": Object {
            "title": "Foo",
          },
          "key": "foo",
        },
      ],
      "noItemsText": "no items text",
    },
  },
}
`;

exports[`gardener-dashboard configmap alert should render the template w/ \`alert.identifier\` 1`] = `
Object {
  "frontend": Object {
    "alert": Object {
      "identifier": "bar",
      "message": "foo",
      "type": "warning",
    },
  },
}
`;

exports[`gardener-dashboard configmap alert should render the template w/o \`alert.identifier\` 1`] = `
Object {
  "frontend": Object {
    "alert": Object {
      "message": "foo",
      "type": "warning",
    },
  },
}
`;

exports[`gardener-dashboard configmap kubeconfig download should render the template w/ \`oidc.public\` 1`] = `
Object {
  "apiServerCaData": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCllYQnBVMlZ5ZG1WeVEyRT0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQ==",
  "apiServerUrl": "https://api.garden.example.org",
  "logFormat": "text",
  "logLevel": "debug",
  "oidc": Object {
    "ca": "-----BEGIN CERTIFICATE-----
Li4u
-----END CERTIFICATE-----",
    "issuer": "https://identity.garden.example.org",
    "public": Object {
      "clientId": "kube-kubectl",
      "clientSecret": "kube-kubectl-secret",
    },
    "redirect_uri": "https://gardener.garden.example.org/auth/callback",
    "rejectUnauthorized": true,
    "scope": "openid email profile groups audience:server:client_id:dashboard audience:server:client_id:kube-kubectl",
  },
  "port": 8080,
  "readinessProbe": Object {
    "periodSeconds": 10,
  },
}
`;

exports[`gardener-dashboard configmap should render the template w/ defaults values 1`] = `
Object {
  "apiVersion": "v1",
  "kind": "ConfigMap",
  "metadata": Object {
    "labels": Object {
      "app": "gardener-dashboard",
      "chart": "gardener-dashboard-0.2.0",
      "heritage": "Helm",
      "release": "dashboard",
    },
    "name": "gardener-dashboard-configmap",
    "namespace": "garden",
  },
}
`;

exports[`gardener-dashboard configmap should render the template w/ defaults values 2`] = `
Object {
  "apiServerUrl": "https://api.garden.example.org",
  "frontend": Object {
    "defaultHibernationSchedule": Object {
      "development": Array [
        Object {
          "end": "00 08 * * 1,2,3,4,5",
          "start": "00 17 * * 1,2,3,4,5",
        },
      ],
      "evaluation": Array [
        Object {
          "start": "00 17 * * 1,2,3,4,5",
        },
      ],
      "production": null,
    },
    "externalTools": Array [
      Object {
        "icon": "apps",
        "title": "Applications and Services Hub",
        "url": "https://apps.garden.example.org/foo/bar{?namespace,name}",
      },
    ],
    "features": Object {
      "projectTerminalShortcutsEnabled": false,
      "terminalEnabled": false,
    },
    "helpMenuItems": Array [
      Object {
        "icon": "description",
        "title": "Getting Started",
        "url": "https://gardener.cloud/about/",
      },
      Object {
        "icon": "mdi-slack",
        "title": "slack",
        "url": "https://kubernetes.slack.com/messages/gardener",
      },
      Object {
        "icon": "mdi-bug",
        "title": "Issues",
        "url": "https://github.com/gardener/dashboard/issues/",
      },
    ],
    "landingPageUrl": "https://gardener.cloud/",
    "seedCandidateDeterminationStrategy": "SameRegion",
  },
  "logFormat": "text",
  "logLevel": "debug",
  "oidc": Object {
    "ca": "-----BEGIN CERTIFICATE-----
Li4u
-----END CERTIFICATE-----",
    "issuer": "https://identity.garden.example.org",
    "redirect_uri": "https://gardener.garden.example.org/auth/callback",
    "rejectUnauthorized": true,
    "scope": "openid email profile groups audience:server:client_id:dashboard audience:server:client_id:kube-kubectl",
  },
  "port": 8080,
  "readinessProbe": Object {
    "periodSeconds": 10,
  },
  "terminal": Object {
    "bootstrap": Object {
      "disabled": true,
    },
    "container": Object {
      "image": "ops-toolbelt:0.11.0-mod1",
    },
    "garden": Object {
      "operatorCredentials": Object {
        "serviceAccountRef": Object {
          "name": "dashboard-terminal-admin",
          "namespace": "garden",
        },
      },
    },
    "gardenTerminalHost": Object {
      "seedRef": "soil-gcp",
    },
  },
}
`;

exports[`gardener-dashboard configmap terminal shortcuts should render the template 1`] = `
Object {
  "frontend": Object {
    "terminal": Object {
      "shortcuts": Array [
        Object {
          "container": Object {
            "args": Array [
              "a",
              "b",
              "c",
            ],
            "command": Array [
              "command",
            ],
            "image": "repo:tag",
          },
          "description": "description",
          "target": "foo-target",
          "title": "title",
        },
      ],
    },
  },
}
`;

exports[`gardener-dashboard configmap themes should render the template 1`] = `
Object {
  "frontend": Object {
    "themes": Object {
      "dark": Object {
        "main-navigation-title": "grey.darken3",
        "primary": "#ff0000",
      },
      "light": Object {
        "main-navigation-title": "grey.darken3",
        "primary": "#ff0000",
      },
    },
  },
}
`;

exports[`gardener-dashboard configmap themes should render the template with light theme values only 1`] = `
Object {
  "frontend": Object {
    "themes": Object {
      "light": Object {
        "main-navigation-title": "grey.darken3",
        "primary": "#ff0000",
      },
    },
  },
}
`;

exports[`gardener-dashboard configmap tickets should render the template 1`] = `
Object {
  "frontend": Object {
    "ticket": Object {
      "avatarSource": "gravatar",
      "gitHubRepoUrl": "https://github.com/gardener/tickets",
      "hideClustersWithLabels": Array [
        "ignore1",
        "ignore2",
      ],
      "issueDescriptionTemplate": "issue description",
      "newTicketLabels": Array [
        "default-label",
      ],
    },
  },
  "gitHub": Object {
    "apiUrl": "https://github.com/api/v3/",
    "org": "gardener",
    "repository": "tickets",
  },
}
`;

exports[`gardener-dashboard configmap tickets should render the template 2`] = `
Object {
  "apiVersion": "v1",
  "data": Object {
    "authentication.token": "d2ViaG9va0F1dGhlbnRpY2F0aW9uVG9rZW4=",
    "webhookSecret": "d2ViaG9va1NlY3JldA==",
  },
  "kind": "Secret",
  "metadata": Object {
    "labels": Object {
      "app": "gardener-dashboard",
      "chart": "gardener-dashboard-0.2.0",
      "heritage": "Helm",
      "release": "dashboard",
    },
    "name": "gardener-dashboard-github",
    "namespace": "garden",
  },
  "type": "Opaque",
}
`;

exports[`gardener-dashboard configmap unreachable seeds should render the template 1`] = `
Object {
  "unreachableSeeds": Object {
    "matchLabels": Object {
      "seed": "unreachable",
    },
  },
}
`;
